plugins/extractors/bigtable/middleware.go

```diff
@@ -0,0 +1,117 @@
+package bigtable
+
+import (
+	"context"
+	"errors"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/bigtable"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"github.com/goto/meteor/utils"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type metricRecorder struct {
+	baseAttributes []attribute.KeyValue
+}
+
+type AdminClientMW struct {
+	tableDuration  metric.Int64Histogram
+	tablesDuration metric.Int64Histogram
+	next           AdminClient
+	metricRecorder metricRecorder
+}
+
+type InstancesAdminClientMW struct {
+	instancesDuration metric.Int64Histogram
+	next              InstanceAdminClient
+	metricRecorder    metricRecorder
+}
+
+func WithAdminClientMW(next AdminClient, projectID, instanceName string) (AdminClient, error) {
+	meter := otel.Meter("")
+
+	tablesDuration, err := meter.Int64Histogram("meteor.bigtable.client.tables.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	tableDuration, err := meter.Int64Histogram("meteor.bigtable.client.table.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &AdminClientMW{
+		tableDuration:  tableDuration,
+		tablesDuration: tablesDuration,
+		next:           next,
+		metricRecorder: metricRecorder{
+			baseAttributes: []attribute.KeyValue{
+				attribute.String("bt.project_id", projectID),
+				attribute.String("bt.instance_name", instanceName),
+			},
+		},
+	}, nil
+}
+
+func WithInstancesAdminClientMW(next InstanceAdminClient, projectID string) (InstanceAdminClient, error) {
+	meter := otel.Meter("")
+
+	instancesDuration, err := meter.Int64Histogram("meteor.bigtable.client.instances.duration", metric.WithUnit("ms"))
+	if err != nil {
+		return nil, err
+	}
+
+	return &InstancesAdminClientMW{
+		instancesDuration: instancesDuration,
+		next:              next,
+		metricRecorder: metricRecorder{
+			baseAttributes: []attribute.KeyValue{
+				attribute.String("bt.project_id", projectID),
+			},
+		},
+	}, nil
+}
+
+func (o *AdminClientMW) Tables(ctx context.Context) (res []string, err error) {
+	defer o.metricRecorder.recordHistogramMetrics(ctx, o.tablesDuration, err)(time.Now())
```

Comment:
like @sudo-suhas mention, `err` will always nil,
also since it is embedded, you can directly call the method

```suggestion
	defer o.recordHistogramMetrics(ctx, o.tablesDuration, func() error {return err})(time.Now())
```

Reply:
updated